parity/url.rs

```diff
@@ -16,33 +16,70 @@
 
 //! Cross-platform open url in default browser
 
+use std;
+
+#[allow(unused)]
+pub enum Error {
+	ProcessError(std::io::Error),
+	FfiNull(std::ffi::NulError),
+	WindowsShellExecute,
+}
+
+impl From<std::io::Error> for Error {
+	fn from(err: std::io::Error) -> Self {
+		Error::ProcessError(err)
+	}
+}
+
+impl From<std::ffi::NulError> for Error {
+	fn from(err: std::ffi::NulError) -> Self {
+		Error::FfiNull(err)
+	}
+}
+
+impl std::fmt::Display for Error {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		match *self {
+			Error::ProcessError(ref e) => write!(f, "{}", e),
+			Error::FfiNull(ref e) => write!(f, "{}", e),
+			Error::WindowsShellExecute => write!(f, "WindowsShellExecute failed"),
+		}
+	}
+}
+
 #[cfg(windows)]
-pub fn open(url: &str) {
+pub fn open(url: &str) -> Result<(), Error> {
 	use std::ffi::CString;
 	use std::ptr;
 	use winapi::um::shellapi::ShellExecuteA;
 	use winapi::um::winuser::SW_SHOWNORMAL as Normal;
+	use winapi::shared::minwindef::INT;
+
+	const WINDOWS_SHELL_EXECUTE_SUCCESS: i32 = 32;
 
-	unsafe {
+	let h_instance = unsafe {
 		ShellExecuteA(ptr::null_mut(),
-			CString::new("open").unwrap().as_ptr(),
-			CString::new(url.to_owned().replace("\n", "%0A")).unwrap().as_ptr(),
+			CString::new("open")?.as_ptr(),
```

Comment:
Isn't `unwrap` "better" for `CString::new("open")`? It logically shouldn't fail.

Reply:
Yeah, you are right it should not fail unless the `&str` contains `0`!

`let ff = CString::new("\u{0}").unwrap();` will fail!